entry.js

```diff
@@ -54,6 +55,170 @@ const hydrate = async () => {
+  }
+
+  // Returns a flattened object while preserving nested key names
+  //  using '.' as a separator ex.
+  //  {parent: {child: 'hello'}} => {parent.child: 'hello'}
+  function flattenObj(input, parentKey="") {
+    let flattened = {};
+    Object.entries(input).forEach(([key, val]) => {
+      if (typeof val === 'object') {
+        flattened = {
+          ...flattened,
+          ...flattenObj(val, `${parentKey}${key}.`),
+        }
+      } else if (typeof val === 'array') {
+         val.forEach((el, idx) => {
+           flattened = {
+             ...flattened,
+             ...flattenObj(el, `${parentKey}${key}.${idx}.`),
+           }
+         })
+      } else {
+        flattened[`${parentKey}${key}`] = val;
+      }
+    });
+    return flattened;
+  }
+
+  // Returns an array of sibling indicies from document root to the param `node`
+  //  so I can use :nth-child selectors to uniquely identify any element
+  function getDOMNodeAddress(node, address = []) {
+    if (node.nodeType === Node.DOCUMENT_NODE) {
+      return address;
+    }
+
+    // Find the index of this node among it's siblings
+    const siblingIndex = Array.from(node.parentNode.children).indexOf(node);
+    return getDOMNodeAddress(node.parentNode, [siblingIndex, ...address]);
+  }
+
+  // re-calculate the HTML string of the component with random props (no update to the visible page)
+  const fakeProps = {
+    document: {streamOutput: randomize(window._RSS_PROPS_.data.document.streamOutput)},
+    __meta: { // throws error if this doesn't exist (?)
+      manifest: { bundlerManifest: {}},
+    },
+  }
+  const randomizedComponentString = renderComponent(fakeProps);
+  const randomizedComponentDOMNode = new DOMParser().parseFromString(randomizedComponentString, 'text/html');
+  const flatProps = flattenObj(fakeProps);
+
+  // For each randomized prop, search the randomized DOM for where the random string occurs
+  Object.entries(flatProps).forEach(([key, val]) => {
+
+    // Create a filtered Node Iterator which will find locations in the DOM where the random string occurs
+    const nodeIterator = document.createNodeIterator(randomizedComponentDOMNode.body, NodeFilter.SHOW_ALL, (node) => {
+      // The string could be textContent for an element
+      if (node.nodeType === Node.TEXT_NODE && node.textContent === val) {
+        node.cfdUsage = 'textContent';
+        return NodeFilter.FILTER_ACCEPT;
+
+      // The string could be an attribute on an element
+      } else if (node.nodeType === Node.ELEMENT_NODE) {
+        const attrs = node.getAttributeNames();
+        for (const attrName of attrs) {
+          if (node.getAttribute(attrName) === val) {
+            node.cfdUsage = `attribute_${attrName}`;
+            return NodeFilter.FILTER_ACCEPT;
+          }
+        }
+      }
+
+      // We didn't find the random string in this node
+      return NodeFilter.FILTER_REJECT;
+    });
+
+    // Use the iterator to modify the real DOM
```

Comment:
One thing I think we could run into is that if we directly modify the dom like this, future react renders would wipe things out. Mostly seems fine but if we're using react to track whether an element's hovered for example, if you hovered it to try and see the field it would re-render and wipe the annotation. Don't know if there's a way to like freeze all rendering?

Reply:
Addressed in later commit